File: cnn.py
import torch
import torch.nn.functional as F
from torch import nn
from game import move_id
import chess
from input import board_to_tensor
'''class CNN(nn.Module):
    def __init__(self, kernel_size, num_kernel_in_first_layer, num_kernel_in_second_layer, padding):
        super(CNN, self).__init__()
        self.kernel_size = kernel_size
        self.num_kernel_in_first_layer = num_kernel_in_first_layer
        self.num_kernel_in_second_layer = num_kernel_in_second_layer
        self.padding = padding
        self.cnn_one = nn.Conv2d(in_channels = num_kernel_in_first_layer,
                                 out_channels = num_kernel_in_second_layer,
                                 kernel_size = kernel_size,
                                 padding = padding)
        
        self.bn1 = nn.BatchNorm2d(num_kernel_in_second_layer)
        self.dropout1 = nn.Dropout2d(0.3)

        self.cnn_two = nn.Conv2d(in_channels = num_kernel_in_second_layer,
                                 out_channels = num_kernel_in_second_layer * 2,
                                 kernel_size = kernel_size,
                                 padding = padding)
        
        self.bn2 = nn.BatchNorm2d(num_kernel_in_second_layer * 2)
        self.dropout2 = nn.Dropout2d(0.3)

        self.fully_connect_one = nn.Linear(num_kernel_in_second_layer * 2 * 8 * 8,
                                           num_kernel_in_second_layer * 2)
        
        self.bn3 = nn.BatchNorm1d(num_kernel_in_second_layer * 2)

        self.fully_connect_two = nn.Linear(num_kernel_in_second_layer * 2, 1)

    def forward(self, x):
        x = F.relu(self.bn1(self.cnn_one(x)))
        x = self.dropout1(x)
        x = F.relu(self.bn2(self.cnn_two(x)))
        x = self.dropout2(x)
        x = torch.flatten(x, 1)
        x = F.relu(self.bn3(self.fully_connect_one(x)))
        return self.fully_connect_two(x)'''

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

class Resnet(nn.Module):
    def __init__(self, filters):
        super(Resnet, self).__init__()
        self.cnn_one = nn.Conv2d(in_channels = filters,
                                 out_channels = filters,
                                 kernel_size = 3,
                                 padding = 1)
        
        self.bn1 = nn.BatchNorm2d(filters)

        self.cnn_two = nn.Conv2d(in_channels = filters,
                                 out_channels = filters,
                                 kernel_size = 3,
                                 padding = 1)
        
        self.bn2 = nn.BatchNorm2d(filters)

    def forward(self, x):
        identity = x
        x = F.relu(self.bn1(self.cnn_one(x)))
        x = self.bn2(self.cnn_two(x))
        x = x + identity
        return F.relu(x)

class CNN(nn.Module):
    def __init__(self, filters, nums_res_block):
        super(CNN, self).__init__()
        self.cnn_one = nn.Conv2d(in_channels = 13,
                                 out_channels = filters,
                                 kernel_size = 3,
                                 padding = 1)
        self.bn1 = nn.BatchNorm2d(filters)
        self.res_blocks = nn.ModuleList()

        for _ in range(nums_res_block):
            block = Resnet(filters)
            self.res_blocks.append(block)
        
        self.policy_cnn = nn.Conv2d(in_channels = filters,
                                    out_channels = 16,
                                    kernel_size = 1)
        self.policy_bn = nn.BatchNorm2d(16)
        self.policy_fully_connect = nn.Linear(16 * 8 * 8, 3820)
        
        self.net_cnn = nn.Conv2d(in_channels = filters,
                                 out_channels = 8,
                                 kernel_size = 1)
        self.net_bn = nn.BatchNorm2d(8)
        self.net_fully_connect = nn.Linear(8 * 8 * 8, 1)
    
    def forward(self, x):
        x = F.relu(self.bn1(self.cnn_one(x)))
        for block in self.res_blocks:
            x = block(x)
        
        policy = F.relu(self.policy_bn(self.policy_cnn(x)))
        policy = torch.flatten(policy, 1)
        policy = self.policy_fully_connect(policy)

        value = F.relu(self.net_bn(self.net_cnn(x)))
        value = torch.flatten(value, 1)
        value = self.net_fully_connect(value)
        value = F.tanh(value)
        
        return policy, value

def masked_policy(board, network):
    network.eval()
    uci = []
    tensor = board_to_tensor(board).unsqueeze(0).to(device)
    network = network.to(device)
    legal_moves = board.legal_moves
    uci = [move.uci() for move in legal_moves]
    indices = [i for i, item in enumerate(move_id) if item in uci]
    with torch.no_grad():
        unmasked_policy, value = network(tensor)
    unmasked_policy = unmasked_policy.flatten()
    masked_logits = torch.full_like(unmasked_policy, float('-inf'))
    masked_logits[indices] = unmasked_policy[indices]

    probs = F.softmax(masked_logits, dim=0)
    move_probs = list(zip(legal_moves, probs[indices]))
    return move_probs, value.item()

'''board = chess.Board("1BQ1qkBN/R6r/1Q6/6BR/2bN4/4Q1BK/1p6/1bq1R1rb w - - 1 3")

class policy_value_network():
    def __init__(self, model = None):
        self.model = model
        self.network = CNN(512, 10)
        self.network.eval()
    
    def policy_value(self, board):
        act_probs, value = masked_policy(board, self.network)'''

File: Tic-Tac-Toe/cnn.cpp
#include "pch.hpp"

#include "Tic_Tac_Toe.hpp"
#include "cnn.hpp"
#include "input.hpp"

residual_block::residual_block(int64_t channels) {
  conv1 = register_module("conv1", torch::nn::Conv2d(torch::nn::Conv2dOptions(channels, channels, 3).stride(1).padding(1).bias(false)));
  conv2 = register_module("conv2", torch::nn::Conv2d(torch::nn::Conv2dOptions(channels, channels, 3).stride(1).padding(1).bias(false)));
  bn1 = register_module("bn1", torch::nn::BatchNorm2d(channels));
  bn2 = register_module("bn2", torch::nn::BatchNorm2d(channels));
}

torch::Tensor residual_block::forward(torch::Tensor x) {
  torch::Tensor out = torch::relu(bn1(conv1(x)));
  out = bn2(conv2(out));
  out += x;
  return torch::relu(out);
}

neural_network::neural_network(int64_t in_channels, int64_t out_channels, int64_t nums_res_blocks) {
  conv = register_module("conv", torch::nn::Conv2d(torch::nn::Conv2dOptions(in_channels, out_channels, 3).stride(1).padding(1).bias(false)));
  policy_conv = register_module("policy_conv",
                                  torch::nn::Conv2d(torch::nn::Conv2dOptions(out_channels, 8, 1).stride(1).padding(0).bias(false)));
  value_conv = register_module("value_conv",
                                 torch::nn::Conv2d(torch::nn::Conv2dOptions(out_channels, 8, 1).stride(1).padding(0).bias(false)));
  bn = register_module("bn", torch::nn::BatchNorm2d(out_channels));
  policy_bn = register_module("policy_bn", torch::nn::BatchNorm2d(8));
  value_bn = register_module("value_bn", torch::nn::BatchNorm2d(8));
  resnet18 = register_module("resnet18", torch::nn::Sequential());
  for (int i = 0; i < nums_res_blocks; i++) {
    auto block = std::make_shared<residual_block>(out_channels);
    resnet18->push_back(block);
  }
  policy_fc = register_module("policy_fc", torch::nn::Linear(8 * 3 * 3, 9));
  value_fc = register_module("value_fc", torch::nn::Linear(8 * 3 * 3, 1));
}

std::pair<torch::Tensor, torch::Tensor> neural_network::forward(torch::Tensor x) {
  x = torch::relu(bn(conv(x)));
  x = resnet18->forward(x);

  torch::Tensor policy = torch::relu(policy_bn(policy_conv(x)));
  policy = policy.view({policy.size(0), -1});
  policy = policy_fc(policy);

  torch::Tensor value = torch::relu(value_bn(value_conv(x)));
  value = value.view({value.size(0), -1});
  value = torch::tanh(value_fc(value));

  return {policy, value};
}


std::pair<std::vector<move_probability>, float> masked_output(board_history& boards, neural_network& network) {
  network.eval();

  torch::Tensor tensor = boards.to_tensor().unsqueeze(0);

  std::vector<int> indices = Tic_Tac_Toe::Move::get_legal_move(boards.get_history().back());

  auto output = network.forward(tensor);
  torch::Tensor unmasked_policy = output.first;
  torch::Tensor value = output.second;

  unmasked_policy = unmasked_policy.flatten();

  torch::Tensor masked_logits = torch::full_like(unmasked_policy, -std::numeric_limits<float>::infinity());
  for (size_t index : indices) {
    masked_logits[index] = unmasked_policy[index];
  }
  torch::Tensor probs = torch::softmax(masked_logits, -1);
  std::vector<move_probability> move_probs;
  for (size_t i = 0; i < indices.size(); ++i) {
    move_probs.push_back(move_probability{indices[i], probs[indices[i]].item<float>()});
  }
  return std::make_pair(move_probs, value.item<float>());
}
